File: Top_K_Frequent_Elements/solution2.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
    let freqMap = {};
    for (let i = nums.length-1; i >= 0; i--) {
        let key = nums[i];
        if (freqMap[key]) {
            ++freqMap[key];
        } else {
            freqMap[key] = 1;
        }
    }
    
    let freqBucket = [];
    freqBucket.length = nums.length + 1;
    Object.keys(freqMap).forEach(key => {
        let freq = freqMap[key];
        if (!freqBucket[freq]) freqBucket[freq] = [];
        freqBucket[freq].push(+key);
    });
    
    
    let ans = [];
    let i = freqBucket.length - 1;
    while (k > 0) {
        let keys = freqBucket[i];
        if (keys) {
            ans.push(...keys);
            if (k === ans.length) k = 0;
        }
        i--;
    }
    return ans
};

File: Top_K_Frequent_Elements/solution2.py
from collections import Counter

class Solution:
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        L = len(nums)
        freqMap = Counter(nums)
        freqBucket = [None] * (L+1)
        for i, key in enumerate(freqMap):
            freq = freqMap[key]
            if freqBucket[freq] == None: freqBucket[freq] = []
            freqBucket[freq].append(key)
        
        i = L
        ans = []
        while k > 0:
            if freqBucket[i] != None: 
                ans += freqBucket[i]
                k -= len(freqBucket[i])
            i -= 1
        return ans
